date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/RecurringTimer.kt

```diff
@@ -0,0 +1,190 @@
+) : HotStateFlowRepo<TimerStateRepo.State>(
+    coroutineScope.coroutineContext,
+    {
+        State.NotRunning.Paused(elapsedSoFar = Duration.ZERO, totalDuration = totalDuration)
+    }
+) {
+    suspend fun start() {
+        withContext(coroutineScope.coroutineContext) {
+            takeAndChangeState { state ->
+                when (state) {
+                    is State.NotRunning.Paused -> suspend {
+                        state.start(interval, coroutineScope, ::finish)
+                    }
+                    is State.NotRunning.Finished, is State.Running -> state.remain()
+                }
+            }
+        }
+    }
+
+    suspend fun pause() {
+        withContext(coroutineScope.coroutineContext) {
+            takeAndChangeState { state ->
+                when (state) {
+                    is State.Running -> state::pause
+                    is State.NotRunning -> state.remain()
+                }
+            }
+        }
+    }
+
+    private suspend fun finish() {
+        withContext(coroutineScope.coroutineContext) {
+            takeAndChangeState { state ->
+                when (state) {
+                    is State.NotRunning.Finished -> state.remain()
+                    is State.Running -> state::finish
+                    is State.NotRunning.Paused -> state::finish
+                }
+            }
+        }
+    }
+
+    /** Timer state. */
+    sealed class State : KalugaState(), Timer.State {
+        abstract val totalDuration: Duration
+        /** Timer is not running. */
+        sealed class NotRunning(override val elapsed: Duration) : State(), Timer.State.NotRunning {
+            /** Timer is paused. */
+            class Paused(
+                elapsedSoFar: Duration,
+                override val totalDuration: Duration
+            ) : NotRunning(elapsedSoFar), Timer.State.NotRunning.Paused {
+                fun start(
+                    interval: Duration,
+                    coroutineScope: CoroutineScope,
+                    finishCallback: suspend () -> Unit
+                ): Running = Running(
+                    elapsedSoFar = elapsed,
+                    totalDuration = totalDuration,
+                    interval = interval,
+                    coroutineScope = coroutineScope,
+                    finishCallback = finishCallback
+                )
+                internal fun finish(): Finished = Finished(totalDuration)
+            }
+
+            /** Timer is finished. */
+            class Finished(
+                override val totalDuration: Duration
+            ) : NotRunning(totalDuration), Timer.State.NotRunning.Finished
+        }
+
+        /** Timer is running. */
+        class Running(
+            elapsedSoFar: Duration,
+            override val totalDuration: Duration,
+            interval: Duration,
+            private val coroutineScope: CoroutineScope,
+            private val finishCallback: suspend () -> Unit
```

Comment:
instead of a callback users can just have a `suspend fun waitForFinish()` which uses a Flow and filter, if they want it async they can put it in a `launch` or `async`. Maybe removing this can simplify some things in this implementation.

Reply:
and `awaitFinish` already exists even.

Reply:
This callback is internally used to set the stateMachine itself to a finished state. So I think this is fine.

Reply:
@thoutbeckers : @Daeda88 is correct, it's a callback to trigger the transition to the finished state. is it possible to improve?

I was thinking on having a SendChannel and send a command, but it seems like an overkill

Reply:
Ok.